### DISCUSSION TITLE
A question about vfio

### DISCUSSION THREAD

**Discussion by zsdaka** [cloud-hypervisor/cloud-hypervisor#4995] [Q&A]
I'm learning the vfio code of cloud hypervisor， I found that only the ept mapping was created for the region with VFIO_REGION_INFO_FLAG_MMAP flag, but no iommu dma mapping was created. Why do these regions not create iommu dma mapping？

**Comment by sboeuf** [comment]
Because these regions represent PCI BARs, which is different from the memory that is used for DMA accesses. There's no DMA transfers through the PCI BARs which is why it's handled differently and doesn't need to be part of the IOMMU on the host. Usually, the IOMMU contains all the mappings corresponding to the entire guest RAM.

**Comment by zsdaka** [comment]
@sboeuf Thank you for your detailed information. If there are two PCI devices A and B, can device A DMA to the bar of device B directly? If so, does it have to establish the iommu mapping of the bar space ？

**Reply by sboeuf** [reply]
I think the main point is that DMA is performed from/to RAM, which means the DMA engine would perform a transfer between the device and the RAM (or vice versa). The PCI BARs are more like a control plane, which allows the guest to program the device.

**Reply by zsdaka** [reply]
@sboeuf Thanks, I got it.

**Comment by zsdaka** [comment]
I have another question about “vfio intx interrupt” that confuses me, cloud hypervisor does not use resamplefd to auto-unmask the interrupt, so bar read/write will be used to detect INTx EOI. But all bar region(except msix-table and PBA) are mapped in add_vfio_device() , so read/write bar spaces will not be trapped. I'm curious whether these codes (INTx EOI in read/write bar) will be executed. Maybe I missed some key program source codes. Please give me some hints, thank you!

**Reply by sboeuf** [reply]
After taking the time to analyze the code, I think you're right and we might have the wrong behavior in that case. I agree that when BARs are entirely or almost (because of MSI-X table and PBA table) entirely mapped, the read/write_bar functions aren't gonna be invoked much (if not at all), meaning we might never acknowledge the INTx by unmasking it. And that would lead to an indefinitely masked IRQ.

We should definitely rely on irqfd resampling as it would be completely transparent for the VMM to unmask the INTx for VFIO. Basically we would create and register a resamplefd, and then we would invoke `VFIO_IRQ_SET_DATA_EVENTFD | VFIO_IRQ_SET_ACTION_UNMASK`. This way, whenever an event is received on the resamplefd, this would automatically trigger the unmasking of the associated INTx.
That would require:
- the `vfio-ioctls` crate to be updated first as we must be able to register an eventfd associated with the unmasking action.
- the use of `register_irqfd_with_resample` from `kvm-ioctls` crate, by creating the associated function in the `hypervisor` crate from Cloud Hypervisor
- patching Cloud Hypervisor codebase to get rid of the BAR read/write deassertion mechanism and replace it with the resampling code

**Reply by sboeuf** [reply]
@zsdaka I've opened an issue for that, do you think you would have some time to take care of it?

**Reply by zsdaka** [reply]
@sboeuf OK, I can have a try.
